### ISSUE TITLE
Exception in device=TPU:3: Input contains NaN, infinity or a value too large for dtype('float32').

### ISSUE THREAD

**Issue by mobassir94** [pytorch/xla#2789]
i tried to convert this pytorch gpu notebook : https://www.kaggle.com/underwearfitting/single-fold-training-of-resnet200d-lb0-965 

into torch xla, my train function and valid function looks like this for tpu  : 

   
```
def macro_multilabel_auc(label, pred):
    aucs = []
    for i in range(len(target_cols)):
        aucs.append(roc_auc_score(label[:, i], pred[:, i]))
    xm.master_print(np.round(aucs, 4))
    return np.mean(aucs)


def train_func(model,device,train_loader):

    model.train()
    bar = tqdm(train_loader)
  
    if use_amp:
        scaler = torch.cuda.amp.GradScaler()
    losses = []
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        
        if use_amp:
            with torch.cuda.amp.autocast():
                logits = model(images)
                loss = criterion(logits, targets)
            scaler.scale(loss).backward()
            scaler.step(optimizer)
            scaler.update()
            optimizer.zero_grad()
        else:
            logits = model(images)
            loss = criterion(logits, targets)
            loss.backward()
            #optimizer.step()
            xm.optimizer_step(optimizer)
            optimizer.zero_grad()
            

        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])

        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')
        #xm.master_print("train_loss = ",train_loss)

    loss_train = np.mean(losses)
    return loss_train


def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS += [logits.sigmoid().detach().cpu().numpy()]
        TARGETS += [targets.detach().cpu()]
        loss = criterion(logits, targets)
        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])
        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')

    PREDS = np.concatenate(PREDS)
    TARGETS = np.concatenate(TARGETS)
    #roc_auc = roc_auc_score(TARGETS.reshape(-1), PREDS.reshape(-1))
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    
    #acc = (image_preds_all==image_targets_all).mean()
    roc_auc = xm.mesh_reduce('test_accuracy', roc_auc, np.mean)
    
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc
```

i  have 2 questions now
1. is my macro_multilabel_auc() implementation here okay for 8 core tpu training?(it worked well on single gpu)
2. why i am getting this error : Exception in device=TPU:3: Input contains NaN, infinity or a value too large for dtype('float32'). ?
full error log : 

```
Exception in device=TPU:3: Input contains NaN, infinity or a value too large for dtype('float32').
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "<timed exec>", line 4, in _mp_fn
  File "<ipython-input-16-b70cadf810f6>", line 86, in train_model
    loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
  File "<ipython-input-15-fbc1b15c4320>", line 81, in valid_func
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
  File "<ipython-input-15-fbc1b15c4320>", line 13, in macro_multilabel_auc
    aucs.append(roc_auc_score(label[:, i], pred[:, i]))
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_ranking.py", line 372, in roc_auc_score
    y_score = check_array(y_score, ensure_2d=False)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 645, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 99, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float32').
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    393         join=join,
    394         daemon=daemon,
--> 395         start_method=start_method)
    396 
    397 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    155 
    156     # Loop on join until it returns True or raises an exception.
--> 157     while not context.join():
    158         pass
    159 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    110                 raise Exception(
    111                     "process %d terminated with exit code %d" %
--> 112                     (error_index, exitcode)
    113                 )
    114 

Exception: process 3 terminated with exit code 17
```

i had Total training sample =  4829 Total validation sample =  1188

the error comes after 81% completion of validation phase

and i used the configuration below : 

fold_id = 0

image_size = 320
seed = 42
warmup_epo = 1
init_lr = 1e-5/3
batch_size = 2 # 64
valid_batch_size = 2
n_epochs = 1
warmup_factor = 10
num_workers = 4

use_amp = False
debug = True # change this to run on full data
early_stop = 5

i didn't observe any NaN value during training on gpu, @taylanbil seeking for your help, thank you

**Comment by mobassir94** [comment]
with the following updated code : 

```

def macro_multilabel_auc(label, pred):
    aucs = []
    for i in range(len(target_cols)):
        aucs.append(roc_auc_score(label[:, i], pred[:, i]))
    
    roc_auc = xm.mesh_reduce('test_accuracy', aucs, np.mean)
    xm.master_print(np.round(roc_auc, 4))
    return roc_auc


def train_func(model,device,train_loader):

    model.train()
    bar = tqdm(train_loader)
  
    if use_amp:
        scaler = torch.cuda.amp.GradScaler()
    losses = []
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        
        if use_amp:
            with torch.cuda.amp.autocast():
                logits = model(images)
                loss = criterion(logits, targets)
            scaler.scale(loss).backward()
            scaler.step(optimizer)
            scaler.update()
            optimizer.zero_grad()
        else:
            logits = model(images)
            loss = criterion(logits, targets)
            loss.backward()
            #optimizer.step()
            xm.optimizer_step(optimizer)
            optimizer.zero_grad()
            

        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])

        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')
        #xm.master_print("train_loss = ",train_loss)

    loss_train = np.mean(losses)
    return loss_train


def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS += [logits.sigmoid().detach().cpu().numpy()]
        TARGETS += [targets.detach().cpu()]
        loss = criterion(logits, targets)
        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])
        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')

    PREDS = np.concatenate(PREDS)
    TARGETS = np.concatenate(TARGETS)
    #roc_auc = roc_auc_score(TARGETS.reshape(-1), PREDS.reshape(-1))
    
    
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    
    #acc = (image_preds_all==image_targets_all).mean()
    
    #roc_auc = xm.mesh_reduce('test_accuracy', roc_auc, np.mean)
    
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc
```

i am getting the same error but it comes after 100% in validation phase : 

```
Exception in device=TPU:0: Input contains NaN, infinity or a value too large for dtype('float32').
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "<timed exec>", line 4, in _mp_fn
  File "<ipython-input-16-b70cadf810f6>", line 86, in train_model
    loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
  File "<ipython-input-15-efaa029dad5a>", line 85, in valid_func
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
  File "<ipython-input-15-efaa029dad5a>", line 13, in macro_multilabel_auc
    aucs.append(roc_auc_score(label[:, i], pred[:, i]))
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_ranking.py", line 372, in roc_auc_score
    y_score = check_array(y_score, ensure_2d=False)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 645, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 99, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float32').
Exception in device=TPU:6: Input contains NaN, infinity or a value too large for dtype('float32').
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "<timed exec>", line 4, in _mp_fn
  File "<ipython-input-16-b70cadf810f6>", line 86, in train_model
    loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
  File "<ipython-input-15-efaa029dad5a>", line 85, in valid_func
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
Exception in device=TPU:3: Input contains NaN, infinity or a value too large for dtype('float32').
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "<ipython-input-15-efaa029dad5a>", line 13, in macro_multilabel_auc
    aucs.append(roc_auc_score(label[:, i], pred[:, i]))
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "<timed exec>", line 4, in _mp_fn
  File "/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_ranking.py", line 372, in roc_auc_score
    y_score = check_array(y_score, ensure_2d=False)
  File "<ipython-input-16-b70cadf810f6>", line 86, in train_model
    loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "<ipython-input-15-efaa029dad5a>", line 85, in valid_func
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
  File "<ipython-input-15-efaa029dad5a>", line 13, in macro_multilabel_auc
    aucs.append(roc_auc_score(label[:, i], pred[:, i]))
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 645, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_ranking.py", line 372, in roc_auc_score
    y_score = check_array(y_score, ensure_2d=False)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 99, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float32').
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 645, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 99, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float32').
Exception in device=TPU:5: Input contains NaN, infinity or a value too large for dtype('float32').
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "<timed exec>", line 4, in _mp_fn
  File "<ipython-input-16-b70cadf810f6>", line 86, in train_model
    loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
  File "<ipython-input-15-efaa029dad5a>", line 85, in valid_func
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
  File "<ipython-input-15-efaa029dad5a>", line 13, in macro_multilabel_auc
    aucs.append(roc_auc_score(label[:, i], pred[:, i]))
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_ranking.py", line 372, in roc_auc_score
    y_score = check_array(y_score, ensure_2d=False)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 645, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 99, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float32').
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    393         join=join,
    394         daemon=daemon,
--> 395         start_method=start_method)
    396 
    397 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    155 
    156     # Loop on join until it returns True or raises an exception.
--> 157     while not context.join():
    158         pass
    159 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    110                 raise Exception(
    111                     "process %d terminated with exit code %d" %
--> 112                     (error_index, exitcode)
    113                 )
    114 

Exception: process 0 terminated with exit code 17
```

**Comment by mobassir94** [comment]
here : https://www.kaggle.com/c/ranzcr-clip-catheter-line-classification/discussion/218299 i can see from first comment that for distributed training,gpu code for multilabel auc is like this : 
```
import torch
import torch.distributed as dist

def gather_tensor(tensor):
    world_size = dist.get_world_size()
    if world_size == 1:
        return tensor
    tensor_list = [torch.zeros_like(tensor) for _ in range(world_size)]
    dist.all_gather(tensor_list, tensor)
    return torch.cat(tensor_list)
```
i think my above multilabel auc implementation is wrong for TPU,can you please suggest me updated multilabel AUC implementation for 8 core torch xla trainingÉ thank you

**Comment by mobassir94** [comment]
is the below implementation correct for TPU? :  


```
def gather_tensor(tensor):
    world_size = xm.xrt_world_size()
    if world_size == 1:
        return tensor
    tensor_list = [torch.zeros_like(tensor) for _ in range(world_size)]
    dist.all_gather(tensor_list, tensor)
    return torch.cat(tensor_list)
```

    
```
def macro_multilabel_auc(label, pred):
    aucs = []
    for i in range(len(target_cols)):
        pred = gather_tensor(pred[:, i])
        label = gather_tensor(label[:, i])
        aucs.append(roc_auc_score(label[:, i], pred[:, i]))
    
    roc_auc = xm.mesh_reduce('test_accuracy', aucs, np.mean)
    xm.master_print(np.round(roc_auc, 4))
    return roc_auc
```

**Comment by taylanbil** [comment]
you can use xm.all_gather to gather remote tensors.

once you gather tensors, you can xm.mark_step, and then send tensors to cpu. after that, you can compute roc however you wish. You will probably have to reshape the gathered tensors first and then send to numpy, qnd then use sklearn.metrics.rocaucscore with binarized labels.

I can work on a prototype next Monday if this doesn't work.

**Comment by mobassir94** [comment]
with the code shown above i got this error Exception in device=TPU:2: zeros_like(): argument 'input' (position 1) must be Tensor, not numpy.ndarray : 

`Exception in device=TPU:2: zeros_like(): argument 'input' (position 1) must be Tensor, not numpy.ndarray
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "<timed exec>", line 4, in _mp_fn
  File "<ipython-input-17-b70cadf810f6>", line 86, in train_model
    loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
  File "<ipython-input-16-95d63a178a40>", line 87, in valid_func
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
  File "<ipython-input-16-95d63a178a40>", line 13, in macro_multilabel_auc
    pred = gather_tensor(pred[:, i])
  File "<ipython-input-15-a0c2f10b6ac6>", line 5, in gather_tensor
    tensor_list = [torch.zeros_like(tensor) for _ in range(world_size)]
  File "<ipython-input-15-a0c2f10b6ac6>", line 5, in <listcomp>
    tensor_list = [torch.zeros_like(tensor) for _ in range(world_size)]
TypeError: zeros_like(): argument 'input' (position 1) must be Tensor, not numpy.ndarray
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    393         join=join,
    394         daemon=daemon,
--> 395         start_method=start_method)
    396 
    397 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    155 
    156     # Loop on join until it returns True or raises an exception.
--> 157     while not context.join():
    158         pass
    159 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    110                 raise Exception(
    111                     "process %d terminated with exit code %d" %
--> 112                     (error_index, exitcode)
    113                 )
    114 

Exception: process 2 terminated with exit code 17`

@taylanbil is mark_tensor necessary? how do i use it? 

is this implementation correct now? i will try it now then : 


```
def gather_tensor(tensor):
    world_size = xm.xrt_world_size()
    if world_size == 1:
        return tensor
    tensor_list = [torch.zeros_like(tensor) for _ in range(world_size)]
    xm.all_gather(tensor_list, tensor)
    return torch.cat(tensor_list).cpu().numpy()
```

    
```
def macro_multilabel_auc(label, pred):
    aucs = []
    for i in range(len(target_cols)):
        pred = gather_tensor(pred[:, i])
        label = gather_tensor(label[:, i])
        aucs.append(roc_auc_score(label[:, i], pred[:, i]))
    
    roc_auc = xm.mesh_reduce('test_accuracy', aucs, np.mean)
    xm.master_print(np.round(roc_auc, 4))
    return roc_auc
```

**Comment by mobassir94** [comment]
i got this error when i tried to use xm.all_gather : 

Exception in device=TPU:0: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "<timed exec>", line 4, in _mp_fn
  File "<ipython-input-17-b70cadf810f6>", line 86, in train_model
    loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
  File "<ipython-input-16-95d63a178a40>", line 87, in valid_func
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
  File "<ipython-input-16-95d63a178a40>", line 13, in macro_multilabel_auc
    pred = gather_tensor(pred[:, i])
  File "<ipython-input-15-331e094cb47d>", line 6, in gather_tensor
    xm.all_gather(tensor_list, tensor)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 596, in all_gather
    if dim < 0:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    393         join=join,
    394         daemon=daemon,
--> 395         start_method=start_method)
    396 
    397 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    155 
    156     # Loop on join until it returns True or raises an exception.
--> 157     while not context.join():
    158         pass
    159 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    110                 raise Exception(
    111                     "process %d terminated with exit code %d" %
--> 112                     (error_index, exitcode)
    113                 )
    114 

Exception: process 0 terminated with exit code 17

please check this on monday @taylanbil 
it will be highly appreciated,thank you

**Comment by mobassir94** [comment]
if this valid_function implementation is correct on GPU : 

```
def valid_func(valid_loader):
    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    with torch.no_grad():
        for batch_idx, (images, targets) in enumerate(bar):

            images, targets = images.to(device), targets.to(device)
            logits = model(images)
            PREDS += [logits.sigmoid()]
            TARGETS += [targets.detach().cpu()]
            loss = criterion(logits, targets)
            losses.append(loss.item())
            smooth_loss = np.mean(losses[-30:])
            bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')
            
    PREDS = torch.cat(PREDS).cpu().numpy()
    TARGETS = torch.cat(TARGETS).cpu().numpy()
    #roc_auc = roc_auc_score(TARGETS.reshape(-1), PREDS.reshape(-1))
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    return loss_valid, roc_auc
```

then is this torch xla conversion correct like above gpu implementation? : 

```
def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS += [logits.sigmoid().detach().cpu().numpy()]
        TARGETS += [targets.detach().cpu()]
        loss = criterion(logits, targets)
        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])
        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')

    PREDS = np.concatenate(PREDS)
    TARGETS = np.concatenate(TARGETS)
    
    
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc
```

**Comment by taylanbil** [comment]
I would implement the roc_auc_score computation as follows:

```python
import numpy as np
import torch
import torch.nn as nn
import torch_xla.core.xla_model as xm
import torch_xla.distributed.xla_multiprocessing as xmp
from sklearn.metrics import roc_auc_score


def macro_multilabel_auc(label, pred):
    aucs = []
    for i in range(pred.shape[-1]):
        aucs.append(roc_auc_score(label[:, i], pred[:, i]))
    xm.master_print('ROCS', np.round(aucs, 4))
    return np.mean(aucs)


def main(*i):
    N, bsz, nclass = 100, 128, 3
    d = xm.xla_device()
    net = nn.Sequential(
        *[
            nn.Linear(N, N),
            nn.Linear(N, nclass),
        ]
    ).to(d)
    inp = torch.randn(bsz, N).to(d)
    labels = torch.randint(low=0, high=2, size=(bsz, nclass)).float().to(d)  # needs to be float for allgather to work
    out = net(inp).sigmoid()
    xm.master_print('SHAPES b4 gather', out.shape, labels.shape)
    out = xm.all_gather(out, dim=0)
    labels = xm.all_gather(labels, dim=0)
    xm.master_print('SHAPES after gather', out.shape, labels.shape)
    # at this point, nothing is executed yet
    # don't send stuff to cpu via .cpu or .item bc of low efficiency.
    xm.mark_step()
    # at this point, out is materialized, you can send to cpu.
    outcpu, labelscpu = out.detach().cpu(), labels.cpu()
    macro_multilabel_auc(labelscpu, outcpu)


if __name__ == '__main__':
    xmp.spawn(main, args=(None,), nprocs=8)
```

you can roll the allgathers inside the roc function as you wish.

The nan-inf-float too large thing is a separate issue, likely not related to your usage. The data has some nan in it likely caused by something unrelated. Could you insert a pdb breakpoint and inspect your tensors (launch 1 process not 8)?

```python
    ...
    xm.mark_step()
    print(torch.isinf(label))
    print(torch.isinf(pred))
    ...
```

**Comment by mobassir94** [comment]
@taylanbil thank you,i was waiting for your reply so badly,,i will try your metric implementation and yes i tried single process instead of 8(IF you check my last comment here carefully) but got same error,
i also observed nan train loss,i don't know why,,,is this implementation correct? : 

```
def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS += [logits.sigmoid().detach().cpu().numpy()]
        TARGETS += [targets.detach().cpu()]
        loss = criterion(logits, targets)
        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])
        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')

    PREDS = np.concatenate(PREDS)
    TARGETS = np.concatenate(TARGETS)
    
    
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc
```

**Comment by mobassir94** [comment]
i was using this : 

```
def macro_multilabel_auc(label, pred):
    aucs = []
    for i in range(len(target_cols)):
        #pred[:, i] = gather_tensor(pred[:, i])
        #label[:, i] = gather_tensor(label[:, i])
        aucs.append(roc_auc_score(label[:, i], pred[:, i]))
    
    #roc_auc = xm.mesh_reduce('test_accuracy', aucs, np.mean)
    roc_auc = np.mean(aucs)
    xm.master_print(np.round(roc_auc, 4))
    return roc_auc
```

then using it in valid function like this : 

```
def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS += [logits.sigmoid().detach().cpu().numpy()]
        TARGETS += [targets.detach().cpu()]
        loss = criterion(logits, targets)
        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])
        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')

    PREDS = np.concatenate(PREDS)
    TARGETS = np.concatenate(TARGETS)
    #roc_auc = roc_auc_score(TARGETS.reshape(-1), PREDS.reshape(-1))
    
    
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    
    #acc = (image_preds_all==image_targets_all).mean()
    
    #roc_auc = xm.mesh_reduce('test_accuracy', roc_auc, np.mean)
    
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc

```

so i can say that the metric implementation that i am trying is correct,right?
only thing was missing there is xm.mark_step()

let me add it,but i am still trying to understand how i am getting nan,,

**Comment by mobassir94** [comment]
i see you are asking me to do pdb breakdown and try : 

print(torch.isinf(label))
print(torch.isinf(pred))

i am confused,why print and not xm.master_print?

**Comment by taylanbil** [comment]
I think nan is not related to your metric implementation. nan is coming from the model probably (this is very likely true since your loss is nan.)

so we need to understand where the nans are coming from.

your metric implementation is inefficient (due to premature .cpu and .item calls). I highly recommend my way of doing it.

Also, I don't think reducing the roc scores themselves makes sense, since it's about the total ranking of scores and 1s and 0s in the labels. Reducing accuracies makes sense, because the total accuracy is the avg of local accuracies. But the total roc score **is not** the average of local roc scores.

**Comment by taylanbil** [comment]
you will need to run in 1 core mode to easily use pdb anyway, so xm.master_print will be same as print.

**Comment by mobassir94** [comment]
@taylanbil thanks
i am trying to update the valid function,i am trying this now  :

```
def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS += [logits.sigmoid()]
        TARGETS += [targets]
        
        xm.master_print('SHAPES b4 gather', PREDS.shape, TARGETS.shape)
        loss = criterion(logits, targets)
        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])
        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')
        
    PREDS = xm.all_gather(PREDS, dim=0)
    TARGETS = xm.all_gather(TARGETS, dim=0)
    
    PREDS = np.concatenate(PREDS)
    TARGETS = np.concatenate(TARGETS)
    
    xm.master_print('SHAPES after gather', PREDS.shape, TARGETS.shape)
    #roc_auc = roc_auc_score(TARGETS.reshape(-1), PREDS.reshape(-1))
    
    # don't send stuff to cpu via .cpu or .item bc of low efficiency.
    xm.mark_step()
    
    PREDS, TARGETS = PREDS.detach().cpu(), TARGETS.cpu()
    
    xm.master_print(torch.isinf(TARGETS))
    xm.master_print(torch.isinf(PREDS))
    
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    
    #acc = (image_preds_all==image_targets_all).mean()
    
    #roc_auc = xm.mesh_reduce('test_accuracy', roc_auc, np.mean)
    
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc

```
is it better now? any bug?

**Comment by taylanbil** [comment]
Here are my recommendations:

- not sure what `bar` is, but it's likely that `bar.set_description` calls .item() underneath and slows you down. I would get rid of it if you don't absolutely need it.
- `bar` doesn't seem to be an instance of `pl.MpDeviceLoader`, which does the sending to device in background, so your code will be more efficient if you use that as in https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py#L232
- there's an `.item()` call in the for loop that slows you down. you can use `losses.append(loss)` instead of loss.item()
- using `+=` to append to lists is less efficient than using .append, you can replace `PREDS += [logits.sigmoid()]` w/ `PREDS.append(logits.sigmoid())`.

These are low hanging fruits in terms of performance improvement. Also, the `nan` definitely comes from one of the iterations inside the loop, so I'd put the breakpoint inside the for loop. somtehing like:

```python

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        xm.mark_step()
        if logits.isinf().any() or logits.isnan().any():
            print('output has inf/nan in step', batch_idx)
            import pdb; pdb.set_trace()

```

**Comment by mobassir94** [comment]
i get this error : 

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    385   pf_cfg = _pre_fork_setup(nprocs)
    386   if pf_cfg.num_devices == 1:
--> 387     _start_fn(0, pf_cfg, fn, args)
    388   else:
    389     return torch.multiprocessing.start_processes(

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in _start_fn(index, pf_cfg, fn, args)
    322   # environment must be fully setup before doing so.
    323   _setup_replication()
--> 324   fn(gindex, *args)
    325 
    326 

<timed exec> in _mp_fn(rank, flags)

<ipython-input-29-b70cadf810f6> in train_model()
     84 
     85         #val_loss, acc = val_epoch(model,device,para_loader.per_device_loader(device))
---> 86         loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
     87 
     88         del para_loader

<ipython-input-28-768c5461d245> in valid_func(model, device, valid_loader)
     85     TARGETS = np.concatenate(TARGETS)
     86 
---> 87     PREDS = xm.all_gather(PREDS, dim=0)
     88     TARGETS = xm.all_gather(TARGETS, dim=0)
     89 

/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py in all_gather(value, dim, groups)
    596   if dim < 0:
    597     dim = value.dim() + dim
--> 598   size = value.size(dim)
    599   padding = [0] * (2 * value.dim())
    600   ordinal = get_ordinal()

TypeError: 'int' object is not callable
```

can you please edit this function of mine? : 

```
def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS += [logits.sigmoid().detach().cpu().numpy()]
        TARGETS += [targets.cpu()]
        
        #xm.master_print('SHAPES b4 gather', PREDS.shape, TARGETS.shape)
        loss = criterion(logits, targets)
        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])
        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')
        
    
    
    PREDS = np.concatenate(PREDS)
    TARGETS = np.concatenate(TARGETS)
    
    PREDS = xm.all_gather(PREDS, dim=0)
    TARGETS = xm.all_gather(TARGETS, dim=0)
    
    xm.master_print('SHAPES after gather', PREDS.shape, TARGETS.shape)
    #roc_auc = roc_auc_score(TARGETS.reshape(-1), PREDS.reshape(-1))
    
    # don't send stuff to cpu via .cpu or .item bc of low efficiency.
    xm.mark_step()
    
    #PREDS, TARGETS = PREDS.detach().cpu(), TARGETS.cpu()
    
    xm.master_print(torch.isinf(TARGETS))
    xm.master_print(torch.isinf(PREDS))
    
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    
    #acc = (image_preds_all==image_targets_all).mean()
    
    #roc_auc = xm.mesh_reduce('test_accuracy', roc_auc, np.mean)
    
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc

```

**Comment by taylanbil** [comment]
`PREDS` and `TARGETS`, the arguments to `xm.all_gather`, have to be tensors, not np.array's. so instead of `np.concatenate`, you need to use something like `torch.cat` (https://pytorch.org/docs/stable/generated/torch.cat.html), which outputs a tensor. Be careful to concatenate on the right dimension.

To address the slowness, please incorporate the suggestions from my previous comment, and then let's see if it's still slow after that.

**Comment by mobassir94** [comment]
should i use xm.mark_step like this? : 

for batch_idx, (images, targets) in enumerate(bar):

    images, targets = images.to(device), targets.to(device)
    
    logits = model(images)
    PREDS += [logits.sigmoid()]
    TARGETS += [targets]
    xm.mark_step()
    if logits.isinf().any() or logits.isnan().any():
        print('output has inf/nan in step', batch_idx)
        import pdb; pdb.set_trace()
    #xm.master_print('SHAPES b4 gather', PREDS.shape, TARGETS.shape)
    loss = criterion(logits, targets)
    losses.append(loss)
    #smooth_loss = np.mean(losses[-30:])
    #bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')

**Comment by mobassir94** [comment]
i get oom during validation phase only,was trying this code : 
```

def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        
        logits = model(images)
        PREDS += [logits.sigmoid()]
        TARGETS += [targets]
    
  
        loss = criterion(logits, targets)
        losses.append(loss)

    
    
    PREDS = torch.cat(PREDS)
    TARGETS = torch.cat(TARGETS)
    
    PREDS = xm.all_gather(PREDS, dim=0)
    TARGETS = xm.all_gather(TARGETS, dim=0)
    
    xm.master_print('SHAPES after gather', PREDS.shape, TARGETS.shape)
    
    # don't send stuff to cpu via .cpu or .item bc of low efficiency.
    xm.mark_step()
    
    PREDS, TARGETS = PREDS.detach().cpu(), TARGETS.cpu()
    
    xm.master_print(torch.isinf(TARGETS))
    xm.master_print(torch.isinf(PREDS))
    
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc

```

error log : 

`---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    385   pf_cfg = _pre_fork_setup(nprocs)
    386   if pf_cfg.num_devices == 1:
--> 387     _start_fn(0, pf_cfg, fn, args)
    388   else:
    389     return torch.multiprocessing.start_processes(

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in _start_fn(index, pf_cfg, fn, args)
    322   # environment must be fully setup before doing so.
    323   _setup_replication()
--> 324   fn(gindex, *args)
    325 
    326 

<timed exec> in _mp_fn(rank, flags)

<ipython-input-19-b70cadf810f6> in train_model()
     84 
     85         #val_loss, acc = val_epoch(model,device,para_loader.per_device_loader(device))
---> 86         loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
     87 
     88         del para_loader

<ipython-input-18-8c0e2601b56f> in valid_func(model, device, valid_loader)
     67     PREDS = []
     68 
---> 69     for batch_idx, (images, targets) in enumerate(bar):
     70 
     71         images, targets = images.to(device), targets.to(device)

/opt/conda/lib/python3.7/site-packages/tqdm/notebook.py in __iter__(self, *args, **kwargs)
    259     def __iter__(self, *args, **kwargs):
    260         try:
--> 261             for obj in super(tqdm_notebook, self).__iter__(*args, **kwargs):
    262                 # return super(tqdm...) will not catch exception
    263                 yield obj

/opt/conda/lib/python3.7/site-packages/tqdm/std.py in __iter__(self)
   1165 
   1166         try:
-> 1167             for obj in iterable:
   1168                 yield obj
   1169                 # Update and possibly print the progressbar.

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/parallel_loader.py in __next__(self)
     29 
     30   def __next__(self):
---> 31     return self.next()
     32 
     33   def __len__(self):

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/parallel_loader.py in next(self)
     35 
     36   def next(self):
---> 37     xm.mark_step()
     38     item = self._loader.next_item(self._device)
     39     if item is None:

/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py in mark_step()
    714   torch_xla._XLAC._xla_step_marker(
    715       torch_xla._XLAC._xla_get_default_device(), [],
--> 716       wait=xu.getenv_as('XLA_SYNC_WAIT', bool, False))
    717   # Only emit metrics from the first local device index, to avoid emitting the
    718   # same values from different threads.

RuntimeError: Resource exhausted: From /job:tpu_worker/replica:0/task:0:
2 root error(s) found.
  (0) Resource exhausted: Attempting to reserve 153.85M at the bottom of memory. That was not possible. There are 100.44M free, 0B reserved, and 100.44M reservable.
	 [[{{node XRTExecute}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

  (1) Resource exhausted: Attempting to reserve 153.85M at the bottom of memory. That was not possible. There are 100.44M free, 0B reserved, and 100.44M reservable.
	 [[{{node XRTExecute}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

	 [[XRTExecute_G10]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

0 successful operations.
0 derived errors ignored.`

**Comment by taylanbil** [comment]
That makes sense, since you're not using `pl.MpDeviceLoader`, there's no `xm.mark_step()` being called, and the graph you end up producing in the for loop becomes huge, and causes the oom when it tries to execute.

In comment https://github.com/pytorch/xla/issues/2789#issuecomment-783675450, I recommended using `pl.MpDeviceLoader`. This takes care of marking step every time it yields a batch. If you don't want to use that, you'll need to add a xm.mark_step() manually at the end of the for loop, so it produces one step's graph and then executes instead of producing N step's graph.

Also I'd definitely switch the `+=`'s to append(), as in `PREDS.append(logits.sigmoid())`

**Comment by mobassir94** [comment]
let me know if this implementation is correct or not : 

i used mark_step

last time i tried it after for loop and got oom,now trying it inside for loop

**Comment by taylanbil** [comment]
that looks ok to me, but I'd add another `xm.mark_step()` just prior to the line w/
`    PREDS, TARGETS = PREDS.detach().cpu(), TARGETS.cpu()`

This way, the allgather actually happens and tensors are materialized.

**Comment by mobassir94** [comment]
i got oom again,,in the middle of validation phase with the code i shared in last comment,, error coming from for loop,please check this : 

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    385   pf_cfg = _pre_fork_setup(nprocs)
    386   if pf_cfg.num_devices == 1:
--> 387     _start_fn(0, pf_cfg, fn, args)
    388   else:
    389     return torch.multiprocessing.start_processes(

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in _start_fn(index, pf_cfg, fn, args)
    322   # environment must be fully setup before doing so.
    323   _setup_replication()
--> 324   fn(gindex, *args)
    325 
    326 

<timed exec> in _mp_fn(rank, flags)

<ipython-input-30-b70cadf810f6> in train_model()
     84 
     85         #val_loss, acc = val_epoch(model,device,para_loader.per_device_loader(device))
---> 86         loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
     87 
     88         del para_loader

<ipython-input-29-a9f0e310b81f> in valid_func(model, device, valid_loader)
     77 
     78         xm.mark_step()
---> 79         if logits.isinf().any() or logits.isnan().any():
     80             print('output has inf/nan in step', batch_idx)
     81             import pdb; pdb.set_trace()

RuntimeError: Resource exhausted: From /job:tpu_worker/replica:0/task:0:
2 root error(s) found.
  (0) Resource exhausted: Attempting to reserve 153.85M at the bottom of memory. That was not possible. There are 104.69M free, 0B reserved, and 104.69M reservable.
	 [[{{node XRTExecute}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

  (1) Resource exhausted: Attempting to reserve 153.85M at the bottom of memory. That was not possible. There are 104.69M free, 0B reserved, and 104.69M reservable.
	 [[{{node XRTExecute}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

	 [[XRTExecute_G10]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

0 successful operations.
0 derived errors ignored.
```

**Comment by taylanbil** [comment]
hmm I wonder if we have to detach before appending to PREDS.. Could you try `PREDS.append(logits.sigmoid().detach())`?

what's the `batch_idx` when you hit this oom?

**Comment by mobassir94** [comment]
tried my past implementation,which was this :  

```
def valid_func(model,device,valid_loader):

    model.eval()
    bar = tqdm(valid_loader)

    PROB = []
    TARGETS = []
    losses = []
    PREDS = []
    
    for batch_idx, (images, targets) in enumerate(bar):

        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS += [logits.sigmoid().detach().cpu().numpy()]
        TARGETS += [targets.detach().cpu()]
        loss = criterion(logits, targets)
        losses.append(loss.item())
        smooth_loss = np.mean(losses[-30:])
        bar.set_description(f'loss: {loss.item():.5f}, smth: {smooth_loss:.5f}')

    PREDS = np.concatenate(PREDS)
    TARGETS = np.concatenate(TARGETS)
    #roc_auc = roc_auc_score(TARGETS.reshape(-1), PREDS.reshape(-1))
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    
    #acc = (image_preds_all==image_targets_all).mean()

    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc
```

it works fine,but you said it is inefficient so i want  your implementation to work,i am trying to add .detach()
i don't know the batch index but according to progress bar i get the error in the middle of validation phase or around middle,,so it is like if there are 300 samples then after 150 or around that number i get OOM

**Comment by mobassir94** [comment]
even after detach i get same OOM,,,after 57% (85/150) step i get OOM

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    385   pf_cfg = _pre_fork_setup(nprocs)
    386   if pf_cfg.num_devices == 1:
--> 387     _start_fn(0, pf_cfg, fn, args)
    388   else:
    389     return torch.multiprocessing.start_processes(

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in _start_fn(index, pf_cfg, fn, args)
    322   # environment must be fully setup before doing so.
    323   _setup_replication()
--> 324   fn(gindex, *args)
    325 
    326 

<timed exec> in _mp_fn(rank, flags)

<ipython-input-26-b70cadf810f6> in train_model()
     84 
     85         #val_loss, acc = val_epoch(model,device,para_loader.per_device_loader(device))
---> 86         loss_valid, roc_auc = valid_func(model,device,para_loader.per_device_loader(device))
     87 
     88         del para_loader

<ipython-input-25-a11bac4f02ee> in valid_func(model, device, valid_loader)
     80 
     81         xm.mark_step()
---> 82         if logits.isinf().any() or logits.isnan().any():
     83             print('output has inf/nan in step', batch_idx)
     84             import pdb; pdb.set_trace()

RuntimeError: Resource exhausted: From /job:tpu_worker/replica:0/task:0:
2 root error(s) found.
  (0) Resource exhausted: Attempting to reserve 153.85M at the bottom of memory. That was not possible. There are 104.69M free, 0B reserved, and 104.69M reservable.
	 [[{{node XRTExecute}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

  (1) Resource exhausted: Attempting to reserve 153.85M at the bottom of memory. That was not possible. There are 104.69M free, 0B reserved, and 104.69M reservable.
	 [[{{node XRTExecute}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

	 [[XRTExecute_G12]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

0 successful operations.
0 derived errors ignored.
```

**Comment by taylanbil** [comment]
ok let's move tensors to cpu then..

based on the code from https://github.com/pytorch/xla/issues/2789#issuecomment-783700227, could you add, after         `xm.mark_step()`
```python
        PREDS.append(logits.sigmoid())
        TARGETS.append(targets)
        xm.mark_step()  # new lines here:
        PREDS[-1] = PREDS[-1].detach().cpu()
        TARGETS[-1] = TARGETS[-1].cpu()

```

**Comment by mobassir94** [comment]
after moving them in CPU i also need to do this after for loop? : 
`PREDS, TARGETS = PREDS.detach().cpu(), TARGETS.cpu()`

**Comment by mobassir94** [comment]
with the code i shared in my last comment of this thread,i again get OOM during validation and this time little bit faster,now i get same OOm after 34% in validation phase

**Comment by zcain117** [comment]
so it sounds like training works with no issues, right?
and validation causes OOM when you try in on TPU, right?

[here](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py#L217) is an example validation loop we recommend for an image model. Note that the [loader here](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py#L233) is wrapped with `pl.MpDeviceLoader`.

Could you try something similar in your case? Start it simple and then try adding things to it until you find the OOM

**Comment by mobassir94** [comment]
@taylanbil and @zcain117 today i tried this implementation : 

```
def valid_func(model,device,valid_loader):
    model.eval()
    bar = tqdm(valid_loader)
    TARGETS = []
    losses = []
    PREDS = []
    for batch_idx, (images, targets) in enumerate(bar):
        images, targets = images.to(device), targets.to(device)
        logits = model(images)
        PREDS.append(logits.sigmoid().detach().cpu().numpy())
        TARGETS.append(targets.detach().cpu())
        loss = criterion(logits, targets)
        losses.append(loss.item())
        xm.mark_step()  # new lines here:
        if logits.isinf().any() or logits.isnan().any():
            xm.master_print('output has inf/nan in step', batch_idx)
            import pdb; pdb.set_trace()
    PREDS = torch.Tensor(np.concatenate(PREDS)).cpu().to(device)
    TARGETS = torch.Tensor(np.concatenate(TARGETS)).cpu().to(device)
    PREDS = xm.all_gather(PREDS, dim=0).cpu().numpy()
    TARGETS = xm.all_gather(TARGETS, dim=0).cpu()
    #xm.master_print('SHAPES after gather', PREDS.shape, TARGETS.shape)
    roc_auc = macro_multilabel_auc(TARGETS, PREDS)
    loss_valid = np.mean(losses)
    xm.master_print("roc_auc = ",roc_auc)
    return loss_valid, roc_auc
```

```
def macro_multilabel_auc(label, pred):
    aucs = []
    for i in range(pred.shape[-1]):
        aucs.append(roc_auc_score(label[:, i], pred[:, i]))
    xm.master_print('ROCS', np.round(aucs, 4))
    return np.mean(aucs)
```

it works fine on tpu(on 8 core training),but i have a question "is this implementation ERROR FREE?"

with GPU within 2 epochs i can get >80% accuracy with ease but  same model on tpu can't even get 50% validation accuracy within 2 epochs, i have a feeling that my implementation maybe not working well on 8 core training,can you please carefully check my implementation? thanks

**Comment by zcain117** [comment]
is the TPU doing better than random?  if so, it seems like it's converging and maybe you need to adjust the hyperparameters to meet the accuracy of your previous model

is the loss curve similar on TPU and GPU?

**Comment by mobassir94** [comment]
loss looks similar but validation accuracy is different.

this is my gpu setting : 

```
fold_id = 0

image_size = 384
seed = 42
warmup_epo = 1
T_0=10
lr=5e-5
min_lr=5e-5

weight_decay = 1e-5

init_lr = 1e-4
batch_size = 2 # 64
valid_batch_size = 2
n_epochs = 80
warmup_factor = 10
num_workers = 0

use_amp = True
debug = False # change this to run on full data
early_stop = 15

optimizer = optim.Adam(model.parameters(), lr=init_lr)
scheduler_warmup = GradualWarmupSchedulerV2(optimizer, multiplier=10, total_epoch=warmup_epo, after_scheduler=scheduler_cosine)

```

TPU setting : 

```
fold_id = 0

image_size = 384
seed = 42
warmup_epo = 1
init_lr = 1e-4/3
batch_size = 4 # 64
valid_batch_size = 4
n_epochs = 20
warmup_factor = 10
num_workers = 4

use_amp = False
debug = False # change this to run on full data
early_stop = 5
kernel_type = 'vit_large_patch16_384' 
modelName = "vit" #normfreenets, vit
net_type = 'vit_large_patch16_384'
optimizer = optim.Adam(model.parameters(), lr=init_lr/warmup_factor)
scheduler_warmup = GradualWarmupSchedulerV2(optimizer, multiplier=10, total_epoch=warmup_epo, after_scheduler=scheduler_cosine)
```

loss function same and it's BCEWithLogits,
on gpu i use batch size 4 and on tpu 8x8 for vit-base
but on gpu training looks like this : 

epoch 1 validation acc : 0.78
epoch 2 validation acc: 0.82

on tpu : 

epoch 1 validation acc : 0.456
epoch 2 validation acc: 0.4599

**Comment by taylanbil** [comment]
- This code is validation only, and there's no accuracy there, is 50% the roc score or accuracy?
- The inf/nan issue seems to be gone, can you confirm?
- I would make the following changes:
  - move `xm.mark_step()` before the line with `PREDS.append(logits.sigmoid().detach().cpu().numpy())`
  - remove `if logits.isinf().any() or logits.isnan().any():` if this is never hit, and roc's are computed fine now.
- not torch_xla related, but you are computing PREDS (and thus roc) off of sigmoid values but loss uses non-sigmoid values, see `loss = criterion(logits, targets)`. Could you verify if this is intended?

**Comment by mobassir94** [comment]
sorry,i mean "roc_auc_max" score

yes ` loss = criterion(logits, targets)` used both on gpu and tpu so for performance large difference it shouldn't be the culprit, it is intended,i can't understand why huge performance gap between gpu and tpu

**Comment by taylanbil** [comment]
roc being around 50% means model is random and it's not learning.

If you're using `nn.BCEWithLogitsLoss`, indeed feeding logits is the correct approach.

I can see that if we didn't apply `.sigmoid()` before feeding to ROC, the array may cointain +/- inf. so that's probably what fixed the original issue.

If loss is the same but roc score is different, then it's likely that there's a bug we are missing in the roc score computation.

**Comment by taylanbil** [comment]
oooh sklearn's roc_auc score takes truth as first arg and predictions as second arg... 😄

**Comment by mobassir94** [comment]
```
def macro_multilabel_auc(label, pred):
    aucs = []
    for i in range(pred.shape[-1]):
        aucs.append(roc_auc_score(label[:, i], pred[:, i]))
    xm.master_print('ROCS', np.round(aucs, 4))
    return np.mean(aucs)
```

this is what i am using so already using truth as first arg and predictions as second arg,no?

**Comment by taylanbil** [comment]
could you print roc's before all gather? print from all cores, do they look significantly better than 50%?

**Comment by mobassir94** [comment]
1 epoch training takes 1 hour on tpu,if i try this in debug mode then it takes less than 10 minutes but then i can't compare it with original gpu trainer model

**Comment by taylanbil** [comment]
your training seems to be suffering from the early `.item()` calls as well. https://github.com/pytorch/xla/pull/2770 addresses this, but to get full speed, I'd strongly advise removing the .item() calls and usage of tqdm.set_description as well. maybe set description every 50 steps or so.

**Comment by taylanbil** [comment]
I think I found the issue. will update soon.

**Comment by mobassir94** [comment]
is colab free tpu slower than kaggle? i think in kaggle my tpu training was much faster

**Comment by taylanbil** [comment]
AFAIK colab uses tpu v2's, I'm not sure what kaggle uses, could be v3's, which would explain speed difference.

**Comment by mobassir94** [comment]
still colab using v2? i thought it's v3 in colab,,yeah kaggle uses v3-8,,it was much faster there :(

**Comment by taylanbil** [comment]
I thought I found the issue but I didn't. I'm back to

- if loss is same b/w gpu and tpu, it's going down etc, then model seems to be converging.
- roc being ~50% suggests model is _not_ converging.
- I cannot spot a mistake in the roc computation code.

**Comment by mobassir94** [comment]
i am training both on gpu and tpu at this moment,,here is a nfnets f2 model with batch size 2 and image size 384 and batch size only 2 looking like : 

```
loss: 0.68432, smth: 0.54912: 100%|██████████| 12040/12040 [1:51:33<00:00,  1.80it/s] 
loss: 0.10789, smth: 0.50917: 100%|██████████| 3002/3002 [09:23<00:00,  5.33it/s]
[0.5357 0.8947 0.9718 0.6164 0.5499 0.9397 0.9338 0.6054 0.5427 0.5188
 0.865 ]
Tue Feb 23 22:52:54 2021 Fold 0, Epoch 1, lr: 0.0000020, loss_train: 0.61211, loss_valid: 0.55333, roc_auc: 0.724904.
roc_auc_max (0.000000 --> 0.724904). Saving model ...
loss: 0.38266, smth: 0.69091: 100%|██████████| 12040/12040 [1:51:17<00:00,  1.80it/s] 
loss: 0.08281, smth: 0.50429: 100%|██████████| 3002/3002 [09:18<00:00,  5.38it/s]
[0.5679 0.9157 0.9806 0.8297 0.8357 0.9541 0.9503 0.5734 0.5249 0.5303
 0.9667]
Wed Feb 24 00:53:34 2021 Fold 0, Epoch 2, lr: 0.0000020, loss_train: 0.55978, loss_valid: 0.53290, roc_auc: 0.784478.
roc_auc_max (0.724904 --> 0.784478). Saving model ...
loss: 0.63453, smth: 0.52372: 100%|██████████| 12040/12040 [1:49:53<00:00,  1.83it/s] 
loss: 0.06106, smth: 0.48750: 100%|██████████| 3002/3002 [09:11<00:00,  5.44it/s]
[0.6105 0.9004 0.9804 0.7398 0.8467 0.956  0.9612 0.6728 0.6153 0.5751
 0.9965]
Wed Feb 24 02:52:43 2021 Fold 0, Epoch 3, lr: 0.0000020, loss_train: 0.54054, loss_valid: 0.50558, roc_auc: 0.804985.
roc_auc_max (0.784478 --> 0.804985). Saving model ...
loss: 0.87506, smth: 0.50161:  39%|███▉      | 4688/12040 [42:48<1:16:44,  1.60it/s]

```

on colab my tpu code is validating,,after few minutes 1 hours training will be done and i can share you the 1st epoch training performance on tpu,,,on tpu i am trying even bigger stuffs,,,on tpu experiment i am using image size 768 batch size 4x8 and dm_nfnet_f0 model

all other hyper parameter setting is similar to GPU,i should expect much better boost in tpu compared to gpu model because i am using largeee image size on tpu,,bigger batch on tpu and all other settings are robust compared to GPU,,let's wait for 1st epoch,almost done

**Comment by mobassir94** [comment]
frustration after 1 hours slow training on colab,,,first epoch done,,and here is the training log again : 

```
Fold =  0
Total training sample =  24080 Total validation sample =  6003
loss: 0.71094, smth: 0.71667: 100%
752/752 [1:57:05<00:00, 9.34s/it]


100%
187/187 [12:33<00:00, 4.03s/it]
100%
187/187 [12:33<00:00, 4.03s/it]

100%
187/187 [11:54<00:00, 3.82s/it]

100%
187/187 [11:02<00:00, 3.54s/it]

100%
187/187 [09:54<00:00, 3.18s/it]

100%
187/187 [09:38<00:00, 3.09s/it]

100%
187/187 [09:06<00:00, 2.92s/it]

100%
187/187 [08:16<00:00, 2.66s/it]








ROCS [0.4485 0.3852 0.6091 0.3457 0.5041 0.281  0.3876 0.484  0.483  0.4849
 0.6485]
roc_auc =  0.46013670446280686
roc_auc_max (0.000000 --> 0.460137). Saving model ...
```
@taylanbil  :(

**Comment by taylanbil** [comment]
can you share colab?

**Comment by mobassir94** [comment]
yes,why not? here it is : https://colab.research.google.com/drive/1oHA1sLe56FVfuJvW_9PBrNVkBWLlUyER?usp=sharing

if you need dataset access for running the notebook please give me your email id then,thanks

**Comment by mobassir94** [comment]
here is the gpu training code that i tried : https://pastebin.com/vN52LU3S

you can compare and see if you can find my mistake that i am doing on tpu,thanks

**Comment by zcain117** [comment]
it looks like TPU training loss is 0.71 vs 0.54 on GPU.  So the loss is much worse on the TPU model, right?  Then it is kind of expected that validation score will be worse too.

You mentioned `all other hyper parameter setting is similar to GPU` - if you are using different batch size and different image size then maybe you need different hyper params. Do you get the same results if you use same image size and same batch size as GPU?

**Comment by mobassir94** [comment]
i can't use same batch size on tpud gpu,,on gpu i use 4 and on tpu i can't use 4,it will be 4x8 or tpu or simply 1x8

and personally i don't think robust tpu setting is depending too much on hyperparameter,on similar hyper parameter i trained models using tpu and on cassava competition and it worked fine there,,,the issue is definitely not related to hyperparameter(i can assure you that), i tried 1e-4/3,1e-5/3 and no matter what hyperparameter i try roc score will be close to 0.46 so it's not the hyperparameter that is causing trouble

**Comment by mobassir94** [comment]
sorry, i think your observation can be right too @zcain117 

probably hyperparameter is causing trouble,,,can you suggest me a good hyperparameter value for my tpu setting? i will try that now then

**Comment by mobassir94** [comment]
now trying learning rate 1e-4 and using scheduler after 1st epoch,just to compare first epoch performance with gpu

previously it was lr = 1e-4/3
and then before starting first epoch,,1 step of gradual warmup scheduler of warmup_factor = 10

now not using scheduler for first epoch and using 1e-4 and adam for first epoch,if it get's low score too then please let me know where i am making mistakes,thanks

**Comment by mobassir94** [comment]
now let's get back to the main problem, @taylanbil loss NaN issue hasn't solved

this model doesn't work on pytorch xla(it gives NaN) : 

```
class RANZCRResNet200D(nn.Module):
    def __init__(self, model_name='resnet200d', out_dim=11, pretrained=False):
        super().__init__()
        self.model = timm.create_model(model_name, pretrained=False)
        n_features = self.model.fc.in_features
        self.model.global_pool = nn.Identity()
        self.model.fc = nn.Identity()
        self.pooling = nn.AdaptiveAvgPool2d(1)
        self.fc = nn.Linear(n_features, out_dim)

    def forward(self, x):
        bs = x.size(0)
        features = self.model(x)
        pooled_features = self.pooling(features).view(bs, -1)
        output = self.fc(pooled_features)
        return output
```
ResNet200D model of came from @rwightman timm library doesn't work on pytorch xla,,,i tried other models and never got NaN but for ResNet200D i got NaN,

**Comment by taylanbil** [comment]
does loss become nan immediately? is there something in this model that's not in another one that works fine?

**Comment by mobassir94** [comment]
loss becomes NaN very fast,i am seeking the paper of this model so that i can learn more, maybe  @rwightman can help

on the other hand @taylanbil i tried different learning rate for adam,like 1e-5 , 1e-4

and without scheduler for first fold,loss is not getting below 0.7

few days ago on another data problem i used wrote 2 torch tpu kernel and never observed a problem like this before, can't understand why train loss is so high even with different high or low learning rate,it worked fine on gpu,,,1 thing that i can tell is my present dataset is extremely inbalance,,it's like 63 positive samples and 20000  negative sample for each class and we have 11 class,,,most of them are highly imbalance like this,,is this a problem for 8 core torch tpu compared to single core GPU?
i used nfnets where we have weight standardization,it shouldn't be the issue IMHO :(

**Comment by mobassir94** [comment]
and new problem,

just got improvement with 1e-6 learning rate and without scheduler but got error xm.save()

check this error log : 

```
SHAPES after gather (288, 11) torch.Size([288, 11])
ROCS [0.8746 0.3652 0.4761 0.3345 0.3626 0.256  0.4737 0.4636 0.5112 0.4485
 0.6362]
roc_auc =  0.4729309203290964
roc_auc_max (0.000000 --> 0.472931). Saving model ...
Exception in device=TPU:5: tensorflow/compiler/xla/xla_client/mesh_service.cc:364 : Failed to meet rendezvous 'torch_xla.core.xla_model.save': Socket closed (14)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "<timed exec>", line 4, in _mp_fn
  File "<ipython-input-19-005c4a3bdd32>", line 109, in train_model
    xm.save(model.state_dict(), f'{model_dir}{kernel_type}_fold{fold_id}_best_AUC.pth')
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 817, in save
    rendezvous('torch_xla.core.xla_model.save')
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 861, in rendezvous
    return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
RuntimeError: tensorflow/compiler/xla/xla_client/mesh_service.cc:364 : Failed to meet rendezvous 'torch_xla.core.xla_model.save': Socket closed (14)
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<timed exec> in <module>

/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    393         join=join,
    394         daemon=daemon,
--> 395         start_method=start_method)
    396 
    397 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    155 
    156     # Loop on join until it returns True or raises an exception.
--> 157     while not context.join():
    158         pass
    159 

/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    110                 raise Exception(
    111                     "process %d terminated with exit code %d" %
--> 112                     (error_index, exitcode)
    113                 )
    114 

Exception: process 5 terminated with exit code 17
```

i hope i will be able to make nfnets working but resnet200d still not working,,even with 1e-6 lr,,,so surely that resnet200d model is not working on torch xla

**Comment by rwightman** [comment]
@mobassir94 not something I can dig into, seems likely to be a task/situation specific tuning issue. The model trains fine in PyTorch GPU. As per docs, you can create any timm model without the classifier head and pooling (or various combos) https://rwightman.github.io/pytorch-image-models/feature_extraction/#create-with-no-classifier-and-pooling .. will be the same for every model (with exception of vit which are a bit diff). And since it just looks like you are changing the head there you can just specify num_classes=11.

A FYI, I now have access to some reasonable TPU resources so I'll be setting up some PyTorch XLA CI for my collection and likely doing some training. Should reduce issues w/ timm.

**Comment by zcain117** [comment]
if you're seeing `rendezvous` error with `xm.save`, make sure you don't have any code that is performing an `if/else` around the checkpoint saving. `xm.save` already makes sure that only 1 of your 8 cores will save the checkpoints, so you don't need anything such as `if master: xm.save()`

**Comment by mobassir94** [comment]
How do i save best weight then?

**Comment by zcain117** [comment]
are your TPU cores all sharing gradients?  If so, every core should have the same (or extremely similar) trained representation. So you would just call `xm.save()` without any if/else surrounding it

For example, after [this line](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py#L171) you could add another line like below:

before:
```
  test_utils.close_summary_writer(writer)
  xm.master_print('Max Accuracy: {:.2f}%'.format(max_accuracy))
  return max_accuracy
```

after:
```
  test_utils.close_summary_writer(writer)
  xm.master_print('Max Accuracy: {:.2f}%'.format(max_accuracy))
  xm.save(model.state_dict(), 'mnist-xm-save.pth')
  return max_accuracy
```
